src/aws/nodemanager/nodemanager.py

```diff
@@ -113,15 +139,36 @@ def generate_heartbeat(self, notify=True):
     def process_heartbeat(self, hb, source) -> Packet:
         hb['source'] = source  # Set the source IP of the heartbeat (i.e., the worker).
         self.notify(hb)  # Forward the heartbeat to the monitor for metrics.
-        if len(self._tasks)>0:  #If there are tasks in the taskpool send a new command to the worker
-            packet=CommandPacket(command="task")
-            current_task=self._tasks.pop(0)
-            packet["task_data"]= current_task.get_task_data()
-            current_task.state=TaskState.RUNNING
-            self._tasks_running.append(current_task)
+
+        if hb['instance_id'] in list(self._task_assignment.keys()) and not hb[
+                                                                               'instance_id'] in list(
+                self._tasks_running.keys()):
```

Comment:
`_tasks_running` is a list but judging from your use of it, you might want to make it a dict. Running the code might save you from these errors.

Reply:
Heh this is not what is written in my code. I will check this again

Reply:
Yea it probably was older code from me.